README.md
# share

A solution for self-hosted sharing of videos and images over the web with moderate privacy-through-obscurity; or, "those with the link can see it".

## Motivation

- Allow grandparents to share videos and photos of grandchildren via email instead of social media
- Control all the data

## Use

1. Manually export media to ./ingest/originals/
2. Run `npm run ingest` from ./
3. Run `npm run build` from ./
4. Copy all files in ./\_site/ to remote server (TODO: automate)

### Prerequisites

#### Install dependencies

The following tools are required for the scripts above to work (Homebrew install commands provided for macOS and Linux):

- [Bash](<https://en.wikipedia.org/wiki/Bash_(Unix_shell)>)
- [exiftool](https://exiftool.org/) (`brew install exiftool`)
- [ffmpeg](https://ffmpeg.org/) (`brew install ffmpeg`)
- [imagemagick](https://imagemagick.org/) (`brew install imagemagick`)
- [jq](https://stedolan.github.io/jq/) (`brew install jq`)
- [s3cmd](https://s3tools.org/s3cmd) (`brew install s3cmd`)
- [Node.js](https://nodejs.org/en) (`brew install node`)

Install the Node.js dependencies to generate the static site by running the following command from the project root (./):

```sh
npm install
```

#### Create .env for secrets

Create a file named `.env` in the project root for storing secrets. This file is gitignored. The following variables are needed:

- `SITE_NAME`
- `SITE_AUTHOR`
- `SITE_DESCRIPTION`
- `CDN_BASE` (URI path with out the published file name, include trailing slash)
- `WWW_BASE` (site url, include trailing slash)
- `BACKUP_PATH` (assumes a local directory connected to cloud storage)
- `S3_BUCKET`

## System overview

- Linux server hosting website (eg [Digital Ocean droplets](https://m.do.co/c/a23535c3f83e))
- S3-compatible data store serving static assets (eg [Digital Ocean Spaces](https://m.do.co/c/a23535c3f83e))
- Manually export videos and images to ./ingest/originals/
- Transform videos and images locally via a script
  - Keep a local and backup CSV record of media metadata (git ignored)
- Generate static site locally via a script
  - One video or image per html page
  - URLs made from media date, type, and shasum
- Copy local site to linux server via a script

## Assumptions

- Media is iOS-centric (.MOV, .JPG)
  - iphone used for taking videos and images so some exif metadata fields might not exist from other sources and file formats might be different
- Certain media metadata fields are recorded (./ingest/fields_of_interest.json)

## Contributing

Open a PR!

## Author

Brian Zelip, https://zelip.me

## License

GNU GPLv3


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

- [ ] add a script for generating an index of all of the media organized by batch
- [x] add to readme the required .env vars
- [ ] add tests
- [x] move the console log at the end of the ingest file to the end of each ingest section, and provide a summary of how many files were ingested.
- [x] console log the links to each page from the ingested media once ingestion is complete

## [1.0.0] - 2023-05-28

### Added

- README.md
- CHANGELOG.md
- LICENSE

## [0.1.0] - 2023-05-27

### Added

```
.
├── _data
│   ├── images.csv (gitignored)
│   ├── site.js
│   └── videos.csv (gitignored)
├── _includes
│   ├── head.html
│   └── styles.css
├── _layouts
│   ├── base.html
│   ├── image.html
│   └── video.html
├── content
│   ├── image.md
│   └── video.md
├── eleventy.config.js
├── ingest
│   ├── fields_of_interest.json
│   ├── ingest.sh
│   ├── originals
│   ├── processed
│   └── wrangle.js
├── package-lock.json
├── package.json
├── public
    └── robots.txt
└── .env (gitignored)
```


ingest/ingest.sh
#!/bin/bash

# Read environment variables
source "../.env"

# Get number of lines in each data file before ingest
img_data_length_old=$(wc -l < ../_data/images.csv)
vid_data_length_old=$(wc -l < ../_data/videos.csv)

# Go to directory with exported original media
cd ./originals

jpg_count=$(ls -1 *.JPG | wc -l)
mp4_count=$(ls -1 *.MP4 | wc -l)

if [ $jpg_count -ge 1 ]; then

    # Write exif data from original images to file
    # -csv - export in csv format, adds SourceFile column
    # -c '%+.6f' - format coordinates as decimal with 6 places and a + or - prefix
    # -ext JPG - .JPG extension
    # . - look in this directory
    # > ../temp_images_og.csv - write to file
    exiftool -csv -c '%+.6f' -ext JPG . > ../temp_images_og.csv

    # Resize and convert images to webp
    magick mogrify -resize 75% -format webp *.JPG

    # Move processed images
    mv ./*.webp ../processed

    # Write exif data from processed images to file
    exiftool -csv -ext webp ../processed > ../temp_images_processed.csv

    # Manually create file for images sha hashes
    echo "FileName,CreateDate,SHA256,SHA256_short" > ../temp_images_shasum.csv

    # The directory prefix in the for loop ahead gets prepended to ${file}
    # but we only want the file name in the store, so cd for no prefix
    cd ../processed

    # Loop over the webp files to clear exif data and write to sha store
    for image in *.webp
    do
        create_date=$(exiftool -CreateDate -T ${image})

        exiftool -all:all= -AllDates=2000:01:01 ${image}

        # shasum returns two strings, the shasum and the filename
        # so isolate the shasum with awk.
        # Trim the sha256 to 15 chars for use in filenames
        hash=$(shasum --algorithm 256 ${image} | awk '{print $1}')
        hash_short=$(echo $hash | cut -c1-15)
        echo $image","$create_date","$hash","$hash_short >> ../temp_images_shasum.csv
    done

    # Go to ingest directory
    cd ../

    # Compute the final metadata for this batch of images and append
    # to _data/images.csv
    node ./wrangle.js "img"

    # Rename processed images using the batch date and their shasum,
    # quicker in bash than node so create a temp json file for jq access.
    # [IFS](https://en.wikipedia.org/wiki/Input_Field_Separators)
    cat temp_images_finalized.json | \
    jq -r '.[] | .Processed_FileName + "," + .Publish_FileName' | \
    while IFS=, read -r old new; do
        mv ./processed/$old ./processed/${new}
    done

    # Copy processed images to CDN, make public, set MIME type, and
    # set cache control to 1 year
    s3cmd put ./processed/*.webp $S3_BUCKET \
    --acl-public \
    --add-header=Cache-Control:max-age=31536000 \
    --no-guess-mime-type \
    --content-type="image/webp"

    # Backup to cloud
    cp ../_data/images.csv "$BACKUP_PATH"

    # Clean up! remove temp files and processed images
    rm ./temp_images_*.{csv,json} ./processed/*.webp*

    # Print success message
    if [ $jpg_count -eq 1 ]; then
        noun="image"
        verb="was"
    else
        noun="images"
        verb="were"
    fi
    echo "🎉 ${jpg_count} ${noun} ${verb} ingested!"
fi

if [ $mp4_count -ge 1 ]; then
    cd ./originals

    exiftool -csv -c '%+.6f' -ext MP4 . > ../temp_videos_og.csv

    # Resize video and save it to ../processed/
    # -n: avoid overwriting output files
    # -loglevel error: show errors and hide the rows of progress
    # -i: input file name
    # -movflags use_metadata_tags: copy metadata from input file
    # -map_metadata 0: copy metadata from input file
    # -vcodec libx264: codec lib to use
    # -crf 30: single-pass compression with minor noticeable difference ("0 = lossless, 23 = default, 51 = worst)
    # -preset faster: faster than default encoding time of 'medium'
    # -tune film: specify input is an HQ video other than 'cartoon', 'stillimage', etc
    for video in *.MP4
    do
        ffmpeg -n -loglevel error -i "$video" \
        -movflags use_metadata_tags \
        -map_metadata 0 \
        -vcodec libx264 \
        -crf 30 \
        -preset faster \
        -tune film \
        "../processed/${video}"
    done

    exiftool -csv -ext MP4 ../processed > ../temp_videos_processed.csv 

    echo "FileName,CreationDate,SHA256,SHA256_short" > ../temp_videos_shasum.csv

    cd ../processed

    for video in *.MP4
    do 
        creation_date=$(exiftool -CreationDate -T ${video})

        exiftool -all:all= -AllDates=2000:01:01 ${video}

        hash=$(shasum --algorithm 256 ${video} | awk '{print $1}')
        hash_short=$(echo $hash | cut -c1-15)
        echo $video","$creation_date","$hash","$hash_short >> ../temp_videos_shasum.csv
    done

    cd ../

    node ./wrangle.js "vid"

    cat temp_videos_finalized.json | \
    jq -r '.[] | .Processed_FileName + "," + .Publish_FileName' | \
    while IFS=, read -r old new; do
        mv ./processed/$old ./processed/${new}
    done

    s3cmd put ./processed/*.mp4 $S3_BUCKET \
    --acl-public \
    --add-header=Cache-Control:max-age=31536000 \
    --no-guess-mime-type \
    --content-type="video/mp4"

    cp ../_data/videos.csv "$BACKUP_PATH"

    rm ./temp_videos_*.{csv,json} ./processed/*.{mp4,MP4}*

    if [ $mp4_count -eq 1 ]; then
        noun="video"
        verb="was"
    else
        noun="videos"
        verb="were"
    fi
    echo "🎉 ${mp4_count} ${noun} ${verb} ingested!"
fi

# Get number of lines in each data file after ingest, print
# published URLs to screen
img_data_length_new=$(wc -l < ../_data/images.csv)
vid_data_length_new=$(wc -l < ../_data/videos.csv)  

if [ $img_data_length_new -gt $img_data_length_old ]; then
    count=1
    echo "📷 New images:"

    for image in $(cat ./_data/images.csv | tail \
    -n $(($img_data_length_new - $img_data_length_old)) | cut \
    -d, -f29)
    do
        echo "${count}. ${image}"
        count=$((count+1))
    done
fi

if [ $vid_data_length_new -gt $vid_data_length_old ]; then
    count=1
    echo "📹 New videos:"

    for video in $(cat ./_data/videos.csv | tail \
    -n $(($vid_data_length_new - $vid_data_length_old)) | cut \
    -d, -f26)
    do
        echo "${count}. ${video}"
        count=$((count+1))
    done
fi


ingest/wrangle.js
const fs = require('fs');
const Papa = require('papaparse');
require('dotenv').config({ path: '../.env' });
const foi = require('./fields_of_interest.json');

const mediaType = process.argv[2]; // 'img' or 'vid'
const mShort = mediaType;
const mLong = mediaType === 'img' ? 'image' : 'video';
const mPlural = mediaType === 'img' ? 'images' : 'videos';

const mediasOriginalCsv = fs.readFileSync(`./temp_${mPlural}_og.csv`, 'utf8');
const mediasProcessedCsv = fs.readFileSync(
  `./temp_${mPlural}_processed.csv`,
  'utf8'
);
const mediasShasumCsv = fs.readFileSync(`./temp_${mPlural}_shasum.csv`, 'utf8');
const mediasFinalizedTempJson = `./temp_${mPlural}_finalized.json`;
const mediasFinalizedCsv = `../_data/${mPlural}.csv`;

const csvReadCfg = { header: true, delimiter: ',', skipEmptyLines: true };
const csvWriteCfg = { header: false, delimiter: ',', skipEmptyLines: true };

const mediasOriginal = Papa.parse(mediasOriginalCsv, csvReadCfg).data;
const mediasProcessed = Papa.parse(mediasProcessedCsv, csvReadCfg).data;
const mediasShasum = Papa.parse(mediasShasumCsv, csvReadCfg).data;

if (
  mediasOriginal.length !== mediasProcessed.length ||
  mediasOriginal.length !== mediasShasum.length
) {
  throw new Error(`The ${mLong} data are not the same length!`);
}

const batchId = batchID(mShort);

const newMediasBase = mediasOriginal.map(
  (med) => new MediaBase(batchId, med, foi[mShort].og)
);

// Add the desired metadata from the processed and shasum csv files
// into the newMediasBase objects created from the originals csv file
const mediasFinalized = newMediasBase.map((med) => {
  // Find the rows in the processed and shasum csv files that correspond
  // to the current row (or media) from the originals csv file
  const proTargetMed = mediasProcessed.find((pMed) => uid(pMed) === med.ID);
  const shaTargetMed = mediasShasum.find((sMed) => uid(sMed) === med.ID);

  if (proTargetMed) {
    addFields(proTargetMed, foi[mShort].processed, med);
  } else {
    throw new Error('No processed target object found!');
  }

  if (shaTargetMed) {
    addFields(shaTargetMed, foi[mShort].shasum, med);
  } else {
    throw new Error('No shashum target object found!');
  }

  // Compute the published ID and its variants from CreateDate or CreationDate,
  // short sha256, and media type. ID ie: 2023-04-06-img-646e8a7b3e7f28f
  const originDate =
    mediaType === 'img' ? med.OG_CreateDate : med.OG_CreationDate;
  const pubYmd = originDate.split(' ')[0].replaceAll(':', '-');
  const pubExt = mediaType === 'img' ? '.webp' : '.mp4';
  const pubId = `${pubYmd}-${mShort}-${med.Processed_SHA256_short}`;
  const pubFileName = `${pubId}${pubExt}`;

  med.Publish_ID = pubId;
  med.Publish_FileName = `${pubFileName}`;
  med.Publish_CDN = `${process.env.CDN_BASE}${pubFileName}`;
  med.Publish_WWW = `${process.env.WWW_BASE}${pubId}.html`;

  return med;
});

// Write mediasFinalized to a temp json file for bash script
try {
  fs.writeFileSync(
    mediasFinalizedTempJson,
    JSON.stringify(mediasFinalized),
    'utf8'
  );
} catch (err) {
  throw new Error(err);
}

// Append mediasFinalized to the appropriate 11ty global data file
try {
  fs.appendFileSync(
    mediasFinalizedCsv,
    Papa.unparse(mediasFinalized, csvWriteCfg) + '\n',
    'utf8'
  );
} catch (err) {
  throw new Error(err);
}

console.log('mediasFinalized', mediasFinalized);

/**
 * MediaBase
 * @constructor
 * @description The starting point for the complete store per media file,
 * comprised of metadata from the original media file and a unique id.
 * The processed and shasum metadata are added after this instantiation.
 * @param {string} batchId - unique id for the batch of files being
 * processed that this media belongs to
 * @param {object} data - an object representing a row from the media
 * CSV file parsed as JSON
 * @param {[string]} fields - array of strings of fields of interest
 */
function MediaBase(batchId, data, fields) {
  this.BatchID = batchId;
  this.ID = uid(data);
  addFields(data, fields, this, 'OG');
}

/**
 * uid
 * @description The unique identifier for a media file, based on the
 * FileName and CreateDate or CreationDate properties.
 * ASSUMES it is being run against csv data not newMediasBase.
 * @param {object} data - a media metadata object from CSV data
 * @returns {string} ie: IMG_3657--2023:03:19
 */
function uid(data) {
  const fileName = data.FileName.split('.')[0];
  const _date = data.FileName.endsWith('.MP4')
    ? data.CreationDate.split(' ')[0]
    : data.CreateDate.split(' ')[0];

  return `${fileName}--${_date}`;
}

/**
 * addFields
 * @description Adds the fields of interest from a csv file as properties on
 * a MediaBase media object
 * @param {object} data - a media metadata object
 * @param {[string]} fields - array of strings of fields of interest
 * @param {object} target - the object to which the properties will be added
 * @param {string} [prefix='Processed'] - prefix for all property names,
 * defaults to 'Processed'
 * @returns {undefined}
 */
function addFields(data, fields, target, prefix = 'Processed') {
  fields.forEach((field) => {
    if (data[field]) {
      target[`${prefix}_${field}`] = data[field];
    }
  });
}

/**
 * batchID
 * @description Creates a filename for the batch of files being processed
 * in the form of YYYY-MM-DD-S-type, where S is the time of day in seconds
 * converted from HH:mm:ss format, and type is the media type. This provides
 * a unique filename with filename-safe characters if multiple batches
 * happen the same day.
 * @param {string} type - the type of media being processed, img or vid
 * @returns {string} - ie: 2023-05-11-66755-img
 */
function batchID(type) {
  const now = new Date().toISOString(); // ie: 2023-05-11T18:28:31.551Z
  const date = now.split('T')[0];
  const seconds = now
    .split('T')[1]
    .split('.')[0]
    .split(':')
    .reduce((acc, cur, i) => {
      return acc + cur * Math.pow(60, 2 - i);
    }, 0);

  return `${date}-${seconds}-${type}`;
}
